SI-6548 reflection now correctly enters jinners

xeno-by opened the pull request:
When completing Java classes, runtime reflection enumerates their
fields, methods, constructors and inner classes, loads them and
enters them into either the instance part (ClassSymbol) or the
static part (ModuleSymbol).

However unlike fields, methods and constructors, inner classes don't
need to be entered explicitly - they are entered implicitly when
being loaded.

This patch fixes the double-enter problem, make sure that enter-on-load
uses the correct owner, and also hardens jclassAsScala against double
enters that can occur in a different scenario.

xeno-by committed:
SI-6548 reflection now correctly enters jinners

When completing Java classes, runtime reflection enumerates their
fields, methods, constructors and inner classes, loads them and
enters them into either the instance part (ClassSymbol) or the
static part (ModuleSymbol).

However unlike fields, methods and constructors, inner classes don't
need to be entered explicitly - they are entered implicitly when
being loaded.

This patch fixes the double-enter problem, make sure that enter-on-load
uses the correct owner, and also hardens jclassAsScala against double
enters that can occur in a different scenario.

Since the fix is about Java-compiled classes, the test needs *.class
artifacts produced by javac. Therefore I updated javac-artifacts.jar to
include the new artifacts along with their source code.

xeno-by commented:
review @retronym @paulp

paulp commented on src/reflect/scala/reflect/runtime/JavaMirrors.scala:
```diff
@@ -1036,14 +1036,14 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
-      val (clazz, module) = createClassModule(owner, name, completer)
-      classCache enter (jclazz, clazz)
-      clazz
+      createClassModule(owner, name, completer) match { case (clazz, module) => clazz }
```
If that method were called createClassAndModule (it's only three characters longer, and far clearer - maybe there is something called a class module! Pretty believable, given that there is something called a module class...) then this could be written

```
createClassAndModule(owner, name, completer)._1
```

and it would be one of those rare uses of tuple selection where it's abundantly clear what's coming out of it.

I need an acronym to mark feedback which is informational/suggestive, not part of my list of hostage demands before I will release little lgtm back to its loving family.  hstc? "here's something to consider!" Anyway, this is of that kind.

xeno-by replied:
That's a very good suggestion! Thanks and done.

paulp commented:
I don't properly understand the code being manipulated here, so I can say "lgtm" but it is a weak lgtm.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1813/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1813/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1103/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1103/

gkossakowski commented on test/files/lib/javac-artifacts.jar.desired.sha1:
```diff
@@ -1 +1 @@
-c5788c5e518eb267445c5a995fd98b2210f90a58 ?javac-artifacts.jar
+61931a51bb1a2d308d214b96d06e9a8808515dcf ?javac-artifacts.jar
```
What this change means? There no mention of in commit message.

xeno-by replied:
There's nothing criminal, but I agree that it's worth mentioning. Updated the message.

xeno-by committed:
cosmetic renamings in runtime reflection

initClassModule => initClassAndModule
createClassModule => initAndEnterClassAndModule

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1105/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1105/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1815/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1815/
